src/main/java/com/zendesk/maxwell/MaxwellUpdateRowsEvent.java

```diff
@@ -48,4 +52,79 @@ public String sqlOperationString() {
 	public String getType() {
 		return "update";
 	}
+
+	private LinkedList<Pair<Row>> filteredRowsBeforeAndAfter;
+	private boolean performedBeforeAndAfterFilter;
+
+	private List<Pair<Row>> filteredRowsBeforeAndAfter() {
+		if ( this.filter == null)
+			return event.getRows();
+
+		if ( performedBeforeAndAfterFilter )
+			return filteredRowsBeforeAndAfter;
+
+		filteredRowsBeforeAndAfter = new LinkedList<>();
+		for ( Pair<Row> p : event.getRows()) {
+			if ( this.filter.matchesRow(this, p.getAfter()) )
+				filteredRowsBeforeAndAfter.add(p);
+		}
+		performedBeforeAndAfterFilter = true;
+		return filteredRowsBeforeAndAfter;
+	}
+
+	@Override
+	public List<RowMap> jsonMaps() {
+		ArrayList<RowMap> list = new ArrayList<>();
+		Object avalue;
+		Object bvalue;
+		for (Iterator<Pair<Row>> p = filteredRowsBeforeAndAfter().iterator(); p.hasNext(); ) {
+			Pair<Row> rowpair = p.next();
+			Row after = rowpair.getAfter();
+			Row before = rowpair.getBefore();
+
+			RowMap rowMap = new RowMap(
+					getType(),
+					getDatabase().getName(),
+					getTable().getName(),
+					getHeader().getTimestamp() / 1000,
+					table.getPKList(),
+					this.getNextBinlogPosition());
+
+			Iterator<Column> aftIter = after.getColumns().iterator();
+			Iterator<Column> befIter = before.getColumns().iterator();
+			Iterator<ColumnDef> defIter = table.getColumnList().iterator();
+			while ( aftIter.hasNext() && defIter.hasNext() && befIter.hasNext() ) {
+				Column a = aftIter.next();
+				ColumnDef d = defIter.next();
+				Column b = befIter.next();
+
+				if ( a instanceof DatetimeColumn) {
+					avalue = ((DatetimeColumn) a).getLongValue();
+				} else {
+					avalue = a.getValue();
+				}
+
+				if ( b instanceof DatetimeColumn) {
+					bvalue = ((DatetimeColumn) b).getLongValue();
+				} else {
+					bvalue = b.getValue();
+				}
+
+				if (avalue != null) {
+					avalue = d.asJSON(avalue);
+					if ( !avalue.equals(d.asJSON(bvalue)) ) {
+						bvalue = bvalue != null ? d.asJSON(bvalue) : bvalue;
+						rowMap.putOldData(d.getName(), bvalue);
```

Comment:
this is a little unclear.  It looks as though you're putting `null` into the `old: {}` section, but so far the convention has been to simply elide NULL fields.  In any case, I think I might write this as spelling out all three cases:

```
   if ( avalue != null && bvalue != null ) { 
       // compare, maybe add to Old
   } else if ( avalue == null && bvalue != null ) {
       //  add to old
   } // else both are null, do nothing.
```

The other possibility to clean up code here is to simply make sure all the fields probably handle `.asJSON(null)`, and then strip null fields later.  But that's a bigger bit of lifting.